README.md
# @bopen/leaflet-area-selection

> Create and manipulate a polygonal area on a Leaflet map

[![NPM](https://img.shields.io/npm/v/@bopen/leaflet-area-selection.svg)](https://www.npmjs.com/package/@bopen/leaflet-area-selection)

[![CI](https://github.com/bopen/leaflet-area-selection/actions/workflows/main.yml/badge.svg)](https://github.com/bopen/leaflet-area-selection/actions/workflows/main.yml)

The primary target of this plugin is to obtain a **customizable selected area** while keeping **high usability**.

## Install

```bash
npm install --save @bopen/leaflet-area-selection
```

## Usage

```javascript
import L from 'leaflet';
import '@bopen/leaflet-area-selection/dist/index.css';
import { DrawAreaSelection } from '@bopen/leaflet-area-selection';

const map = L.map('root').setView([41.901493, 12.5009157], 13);

L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
  attribution:
    '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
}).addTo(map);

const areaSelection = new DrawAreaSelection();

map.addControl(areaSelection);
```

### Old school usage?

Although we think it's time for Leaflet to move away from global `L` object and open to modern JavaScript and tree-shaking, you can still use this plugin in the old way.

This gist contains a modified version of the Leaflet "Hello World" example: https://gist.github.com/keul/74e442b96a41c4f50e304c22259a63c3

## Configuration

The `DrawAreaSelection` can receive a configuration object with followings supported options:

- `active` - the plugin starts as active, no need to run it from the button
- `fadeOnActivation` - when actively drawing a new area the map is partially faded out
- `onPolygonReady` - callback called during the adjustment phase (see below) every time a change is performed to the polygon (vertex moved, added or removed).<br>
  Receives the [Leaflet.Polygon](https://leafletjs.com/reference.html#polygon) and the control instance as arguments.
- `onPolygonDblClick` - called when performing a double-click on the draw polygon when in the adjustment phase (see below).<br>
  Receives three arguments: the [Leaflet.Polygon](https://leafletjs.com/reference.html#polygon), the control instance and the event.
- `onButtonActivate` - called when user clicks on the button to enable draw a polygon.<br>
  Receives two arguments: the control instance and the event. Calling `preventDefault` on the event will abort the activation.
- `onButtonDeactivate` - called when user clicks on the button to stop drawing the polygon.<br>
  Receives three arguments: the [Leaflet.Polygon](https://leafletjs.com/reference.html#polygon), the control instance and the event.
  Calling `preventDefault` on the event will abort the deactivation.

## User guide

### Defining the skeleton of the polygon

When the plugin is activated by using the new control, the map enters in a **drawing phase**.<br>
Clicks on the map will trigger the creation of a vertex for the polygon.

User can continue adding vertexes to the polygon (min length is 3) until the whole required area is covered.<br>
To complete the polygon drawing phase user must click on the first (green) point created.

Alternatively: on the first click you can drag&drop and directly create a square shape (_note_: this method is not currently working on mobile!).

### Fix and change polygon shape

At this point we enter the **adjustment phase**, where user can:

- move edges of the polygon by dragging them
- create new edges, by dragging the ghost markers in the middle of every path
- deleting edges by double clicking on them

## See it in action

![Example animation](./example.gif)

![Example creating a rect](./example-rect.gif)

See also the [live example](https://bopen.github.io/leaflet-area-selection).

## Credits

This plugin is heavily inspired by the "_draw on map_" feature provided on [Immobiliare.it](https://www.immobiliare.it) website.

## License

MIT © [B-Open](https://www.bopen.eu/)

<img alt="B-Open Logo" src="./example/src/B-Open.svg" width="50" />


CHANGELOG.md
## Changelog

### 0.7.1

- Fixed: map drag is restored when a polygon creation has been cancelled.
  See [#11](https://github.com/bopen/leaflet-area-selection/issues/11)
- Fixed: clicking on the initial marker will sometimes break the creation loop.
  See [#12](https://github.com/bopen/leaflet-area-selection/issues/12)

### 0.7.0

- Added: `activate` method, to make plugin activable externally

### 0.6.2

- Bugfix: drag event were called while user was just performing first click
- Bugfix: ghost markers were displayed in a fuzzy position after a zoom change

### 0.6.1

- Added message to help users on how to draw something

### 0.6.0

- Added: square selection during drawing phase
- Fixed: plugin was not working on mobile if [dragging](https://leafletjs.com/reference-1.7.1.html#map-dragging) was disabled

### 0.5.0

- Added: `onButtonActivate`
- Added: `onButtonDeactivate`

### 0.4.0

- Do not ever ever ever use leaflet-src.esm. _Never_.
- Prevent click events on markers from being propagated to the map
- Fixed issue positioning new markers when the map is not fullscreen
- Added a `drawing-area` CSS class to the map container when plugin is active


example/src/index.js
/* eslint-disable no-console */
import { Map, Icon, tileLayer, marker, geoJSON, Browser } from 'leaflet';
import 'leaflet/dist/leaflet.css';
import '@bopen/leaflet-area-selection/dist/index.css';
import { DrawAreaSelection } from '@bopen/leaflet-area-selection';

import './index.css';

import icon from './B-Open.svg';

// See https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
delete Icon.Default.prototype._getIconUrl;
Icon.Default.mergeOptions({
  iconRetinaUrl: require('leaflet/dist/images/marker-icon-2x.png'),
  iconUrl: require('leaflet/dist/images/marker-icon.png'),
  shadowUrl: require('leaflet/dist/images/marker-shadow.png'),
});

const center = [41.901493, 12.5009157];

const map = new Map('root', {
  // Fix for a bug in Leaflet 1.7.1. See https://github.com/Leaflet/Leaflet/issues/7255
  tap: false,
  dragging: !Browser.mobile,
}).setView(center, 13);

tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
  attribution:
    '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
})
  .on('load', () => {
    window.MAP_LOADED = true;
  })
  .addTo(map);

const popup = document.createElement('div');
popup.innerHTML = `<div style="text-align: center; font-size: 120%">B-Open</div>
<img alt="B-Open logo" width="75" src="${icon}" />
`;

marker(center).addTo(map).bindPopup(popup);

const areaSelection = new DrawAreaSelection({
  onPolygonReady: (polygon) => {
    const preview = document.getElementById('polygon');
    preview.textContent = JSON.stringify(polygon.toGeoJSON(3), undefined, 2);
    preview.scrollTop = preview.scrollHeight;
  },
  onPolygonDblClick: (polygon, control, ev) => {
    const geojson = geoJSON(polygon.toGeoJSON(), {
      style: {
        opacity: 0.5,
        fillOpacity: 0.2,
        color: 'red',
      },
    });
    geojson.addTo(map);
    control.deactivate();
  },
  onButtonActivate: () => {
    const preview = document.getElementById('polygon');
    preview.textContent = 'Please, draw your polygon';
    console.log('Please, draw your polygon');
  },
  onButtonDeactivate: (polygon) => {
    const preview = document.getElementById('polygon');
    console.log('Deactivated');
    preview.textContent = `Deactivated! Current polygon is:

${polygon ? JSON.stringify(polygon.toGeoJSON(3), undefined, 2) : 'null'}`;
  },
  position: 'topleft',
});

map.addControl(areaSelection);

const command = document.querySelector('#control');
command.addEventListener('click', function () {
  if (areaSelection.phase !== 'inactive') {
    areaSelection.deactivate();
  } else {
    areaSelection.activate();
  }
});
